### PR TITLE
CI: Run Flatpak Build on GitHub labeled event

### PR DESCRIPTION
**Pull request by gxalpha**
### Description

Makes the "Flatpack (Experimental)" build automatically rerun when the
label "Seeking Testers" is added to a pull request.

The task will be skipped on other labels. The default pull_request types
are kept.

### Motivation and Context

To run the Flatpack Build after adding the "Seeking Testers" label, the workflow had to be rerun, which is an additional two clicks and not very convenient.
Now, when adding the label, the workflow should automatically rerun. 

### How Has This Been Tested?

Committed to my own master branch and started a PR to it, as usual the workflow was mostly skipped.
After adding the label, it was automatically restarted and started building.

### Types of changes

- Performance enhancement (non-breaking change which improves efficiency)

^ I think I'd put it there, not sure to be honest

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
Can you also make this re-run all build steps on seeking testers added? This is an issue for all build pipelines.

**Comment by gxalpha** [comment]
You mean basically apply this change to all workflows?

**Comment by kkartaltepe** [comment]
For the win64 and mac builds basically, yes.

**Comment by GeorgesStavracas** [comment]
Small pedantry: it's spelled Flatpak, without the *c* letter.

File: .github/workflows/flatpak.yml
Comment (GeorgesStavracas) [inline .github/workflows/flatpak.yml:10]: Nitpick: please remove the space before and after the brackets - the rest of the files doesn't do that.

Another nitpick: please sort these alphabetically → `types: [labeled, opened, reopened, synchronize]`

Comment (gxalpha) [reply .github/workflows/flatpak.yml:10]: Sure

**Comment by gxalpha** [comment]
Added in @GeorgesStavracas feedback, including the typo for which I am very ashamed

Regarding to what @kkartaltepe said, I was gonna do that, but didn't because adding *any* label would retrigger the entire thing.
Which actually also applies to the Flatpak build, though that isn't that bad since "Seeking Testers" would be one of the last labels added, and it doesn't run when the label isn't there (making this a non-issue in my eyes); whereas the "CI Multiplatform Build" does, meaning it would rerun on every label add.
While that could be fixed, it would need some more digging into the current workflow from my side.

Also I accidentally added my browser into this commit, gonna force-push again

**Comment by GeorgesStavracas** [comment]
Nothing to be ashamed of, this is a super common mistake :slightly_smiling_face:

**Review by GeorgesStavracas — APPROVED** [review]
Not a GitHub Actions expert, but comparing this to the [online documentation](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request), it seems correct to me.

**Comment by PatTheMav** [comment]
There is a reason why I didn't add the labeling event to workflow triggers and its precisely for the reasons outlined above. We also add labels as house-keeping at later stages, so as long as Github doesn't allow specific labels as triggers, I'm against merging this.

**Comment by gxalpha** [comment]
It actually does allow for specific labels, that would need some more changing though in the part where it checks if the workflow should run.
If requesed, I can try to add that

**Comment by RytoEX** [comment]
I agree with @PatTheMav .  As this is, a run will be triggered for this workflow for each label added.  If it's possible to confine automatic runs to the specific label, then we'd like to see that patch instead.

**Comment by gxalpha** [comment]
So, haven't tested it yet, but this should fix the "any label being applied" problem.

I will run some tests on my own repo, once I confirmed it works I'll report back, but feel free to read over it and check if I missed / screwed up something really obvious

**Comment by gxalpha** [comment]
Alright, ran some tests, fixed all the stupid bugs, should work now

**Comment by gxalpha** [comment]
@RytoEX @PatTheMav what it does now is the following:
It will always "enter" the run, for each push (same as the current behaviour), but now also for each label. Then it will do either of the following:
- On pushes, it will look for the "Seeking Testers" label, and continue if that label exists (same as the current behaviour)
- If a label is added, it will look which label got added. If it is called "Seeking Testers", it will continue, for any other label (say for example "Bug") it'll abort (skip the build process).

Is this what you expected?

This should also easily be applicable to the "CI Multiplatform Build". Should I also add this? If so, should it be added to this PR or should I open a new one for that?

**Comment by PatTheMav** [comment]
I might be wrong, but this change would mean that if a new push triggers a CI run, it will not check for the available labels, and even though the `Seeking Testers` label is set, it will not detect that and thus not build any artifacts.

The check for that specific label needs to happen on every run (which is why we use a separate API call in the first place) so it can pick up a changed label when a workflow run is re-triggered.

Some additional thoughts:
* Workflows are being re-done right now anyway, flatpak becomes part of our "main" workflow
* We add/remove labels all the time, this would lead to lots of workflow runs that did "nothing" (because labeling triggered it, but the label was not "Seeking Testers")
* Only members can add the necessary label and so far members usually re-run the workflow anyway
* If a member forgot about that, you can trigger another run with an empty "trigger CI" commit
* Such re-runs should only occur in the draft stage, a PR out of draft is "ready to merge" and should not require additional runs
* When in draft stage you can do "trigger CI" commits as much as you'd like, you'll probably need to squash before converting from draft anyway
* If bugs are found, a force push is usually necessary (coupled with the need/request to squash commits), which triggers a re-run by itself

Alas I don't see how you can tell GitHub Actions to run only when a specific label is added without running on every label action done in the repo.

**Comment by gxalpha** [comment]
> I might be wrong, but this change would mean that if a new push triggers a CI run, it will not check for the available labels, and even though the Seeking Testers label is set, it will not detect that and thus not build any artifacts.

The push event should still act like it did before, that functionality doesn't get removed.

Reading your additional notes though, I take that this PR can safely be closed?

**Comment by PatTheMav** [comment]
> Reading your additional notes though, I take that this PR can safely be closed?

I'm just one voice of many, I would prefer to hear what others think first.

**Comment by RytoEX** [comment]
> If a label is added, it will look which label got added. If it is called "Seeking Testers", it will continue, for any other label (say for example "Bug") it'll abort (skip the build process).

How exactly does it skip the build process?  As far as I can tell, the `if` in a step just determines if that step runs so long as:
1. the previous step succeeded (didn't fail)
2. the current step's if conditions evaluate to true

In this case, I think both steps will always return "success" if they run, because I don't think "skipped" counts as a failure, so then the checkout and build steps will also run.

https://github.community/t/workflow-skips-steps-besides-if-condition/115956/2
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif
https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions
https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context


You could try to check the step outcome and exit if the "check PR" steps were "skipped" and then exit the job.  It would probably be easier to have a step that checks if the event was "labeled" and the label is not "Seeking Testers" and then exit the job.  Of course, the next thing to figure out would be how to exit the workflow without causing a CI failure.  I'd need to see some test runs of something like that actually working before I'd merge it.

That said, I'm wondering if the fuss is worth it, given @PatTheMav 's points.

**Comment by gxalpha** [comment]
I merged my commit into my master, you should be able to see the results:

When any label (in my case "Bug" is added):
image
(https://github.com/gxalpha/obs-studio/pull/3/checks?check_run_id=2466260484)


When the "Seeking testers" label is added, this is the result:
image
(https://github.com/gxalpha/obs-studio/pull/3/checks?check_run_id=2466317592)

>the current step's if conditions evaluate to true

And that is the reason, since `$SEEKING_TESTERS` is not equal to 1 if not explicitly set.


>I'm wondering if the fuss is worth it

Yeah, I guess that comes down to opinion then.


Also I gave maintainers access to my repo with the checkbox on this PR, are you able to add labels etc or would I have to add you specifically?
If you are able to, feel free to play around and test whatever you want to at
https://github.com/gxalpha/obs-studio/pull/3

**Comment by RytoEX** [comment]
> Also I gave maintainers access to my repo with the checkbox on this PR

Maintainer access is just push/write access to that branch.  It doesn't let us manage labels on PRs.

**Pull request closed without merging**
